src/bot/conversation.py

```diff
@@ -5,79 +5,81 @@
+
 
 load_dotenv()
 openai.api_key = os.getenv("OPENAI_API_KEY")
 client = OpenAI()
-user_id = os.getenv("USER_ID")
-
-all_words = get_all_words(user_id)
-"""extract all words of a dictionary - for now i have it locally"""
-
-
-german_words = []
-"""array of all german words in the dictionary"""
-for i in range(len(all_words)):
-    german_words.append(str(all_words[i][1]))
-german_words = str(german_words)
+chat_history = []
+"""array to store the conversation history locally, initialized with a system message"""
 
 
-userConversation = []
-"""array to store the conversation history locally, initialized with a system message"""
+def initializeConversation(message, bot, newConversation):
+    global user_known_words, chat_history
+    user_id = get_or_create_user_id_in_DB(str(message.from_user.id), message.from_user.username, message.from_user.first_name, message.from_user.last_name)
+    all_words = get_all_words(user_id)
+    user_known_words = []
+    for i in range(len(all_words)):
+        user_known_words.append(str(all_words[i][1]))
+    user_known_words = str(user_known_words) #might be misunderstandable the fact that in random terms i call user_known_words the list of german words + translations while here only german words
+    print(f"\nDictionary from user {message.from_user.first_name} and dictionary is: containing {len(all_words)} words\n")
+    _manageConversation(message, bot, newConversation, [])
 
 
-def callOpenAI(message, bot, newConversation):
+def _manageConversation(message, bot, newConversation, chat_history):
     """
-    Function to start a conversation with the user and get responses from OpenAI
+    Function to start or continue a conversation with the user and get responses from OpenAI
     
     args:
     message: the message object from the user
     bot: the bot object to send the message
     newConversation: a boolean indicating if this is a new conversation or not
+    chat_history: the conversation history array
     
     returns:
-    userConversation: the conversation history
+    chat_history: the conversation history
     """
     try:
         if newConversation:
-            userConversation = []
-            userConversation.append({"role": "system", "content": "You are a bot that helps students to learn German. You need to have simple conversations, with short sentences, using only present tense. You will mainly use terms from the dictionary in the TermsList file, as these are the words the student knows. \nHere is the list of the terms: " + german_words})
+            chat_history = []
+            chat_history.append({"role": "system", "content": f"You are a bot that helps students to learn German. You need to have simple conversations, with short sentences, using only present tense. You will mainly use terms from the dictionary in the TermsList file, as these are the words the student knows. \nHere is the list of the terms: {user_known_words}"})
```

Comment:
Does that work? Where is `user_known_words` coming from? This doesn't seem to be saved on a per-user bases

Reply:
But `user_known_words` is global and you only have ONE instance of this in your program as far as I can see it. So whenever a user initiates a new chat, your entire server only has that users list of words. How can this work, when both users chat at the same time? I don't get it.

But if it works ok. Then Python does some magic here. Otherwise, let's open 2 chats again and investigate.